Basic CLI

a-recknagel opened the pull request:
Closes #16

Commit:
CLI, basic commands and their help texts, and a backend stub

Commit:
Forgot to run black before pushing.

Commit:
Forgot to type check properly

a-recknagel commented:
I'm leaving this open for a while, reviews welcome. If someone is working on a backend/parser module already I'm happy to change my stub to whichever API they chose.

Commit:
Forgot tests as well. Can't believe some people pay me for my 'work'.

- added pyfakefs to test dependencies for temp test files
- changed imports in cli to modules only to make monkey patching easier
- wrote stub results instead of skipped tests in order to get a better
  coverage...

Commit:
Forgot black again.. the pre-commit hook's priority rises by the second

a-recknagel committed:
Add test dependency to CI

Commit:
Added module docstrings.

Initially only for the parser stub to make it clear that it is not
binding, then expanded to other non-test files since it is a nice thing
to do.

maxfischer2781 commented on src/stenotype/cli.py:
```diff
@@ -0,0 +1,82 @@
+
+    # early exits
+    if shorten and check:
+        print("The flags --shorten and --check can not be used at the same time.")
```
Consider writing diagnostics to ``stderr`` -- this allows safely invoking ``stenotype`` by other tools to get the translated types.

a-recknagel replied:
good call

maxfischer2781 commented on src/stenotype/cli.py:
```diff
@@ -0,0 +1,82 @@
+    if shorten:
+        parser.inverse_parse(args, context)
+    else:
+        parser.parse(args, context)
```
The CLI should return a non-zero exit code if parsing does not work (see above comment on use by other tools). Consider defining an Exception for parsing failures, and having the CLI frontend transform that to a human-readable error message.

Perhaps just drop a ``# TODO`` and have this as a good-first-issue.

a-recknagel replied:
These kind of Exceptions are usually just no-content `class StenotypeException(Exception): pass`, right?

I think I'll add that section already, it will be changed once the backend API s written anyway and some of the testing with `click` can be strange.

a-recknagel replied:
> having the CLI frontend transform that to a human-readable error message

Right now it's just `click.echo(e, err=True)`, would you say `traceback.print_exc()` is better?

edit: that was probably what you meant in the first place, but I'll leave finding a good transformation as a good-first-issue =D

maxfischer2781 replied:
We can go through various levels of error report quality -- from a simple ``SyntaxError: "A and B"`` to ``SyntaxError: unrecognised "and" in stenotype "A and B"`` to some [ELM/Rust like ASCII help printout](https://elm-lang.org/news/compiler-errors-for-humans).

So yeah, ``click.echo`` seems like a good start unless someone is really, really bored. ;)

maxfischer2781 reviewed (commented):
Added a few comments for usability by humans/tools.

Commit:
Feedback from CR and some other discussion

- decided to drop "context" option, no custom type assertions with steno
- changed from "print" to "click.echo" with error stream flags to
  improve stability and enabling noisy stderr & structured stdout split
- added custom exceptions to stenotype.util
- added test for multi-input CLI option

Commit:
added test case that asserts custom exception handling behavior

Commit:
._.

Commit:
added autodocs for CLI

Commit:
forgot dependency, again

I don't wanna duplicate package installs over pyproject and CI/CD config
any more, so the pipeline is now kicked off with a dev-install. It
should behave the same for the package that is tested, and makes life a
but easier.

Commit:
updating lockfile

Commit:
running package installs directly within container

Commit:
running package installs directly within container, next try

a-recknagel committed:
running package installs directly within container, next next try

a-recknagel committed:
debugging poetry, need to stop it from creating virtualenvs...

a-recknagel committed:
update export usage to 1.0.0b1 status
